plugin/plugin.go

```diff
@@ -0,0 +1,210 @@
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/ghodss/yaml"
+)
+
+const metadataFileName = "plugin.yaml"
+
+// Command is the command to be executed by conftest,
+// passed as a single string.
+type Command string
+
+// Prepare prepares the plugin command and parses out the
+// main and args
+func (c Command) Prepare() (string, []string, error) {
+	args := strings.Split(os.ExpandEnv(string(c)), " ")
+	if len(args) == 0 || args[0] == "" {
+		return "", nil, fmt.Errorf("prepare plugin command: no command found")
+	}
+
+	main := args[0]
+	var cmdArgs []string
+	if len(args) > 1 {
+		cmdArgs = args[1:]
+	}
+
+	return main, cmdArgs, nil
+}
+
+// MetaData contains the required metadata for conftest plugins
+type MetaData struct {
+	// Name is the name of the plugin
+	Name string `yaml:"name"`
+
+	// Version is a SemVer 2 version of the plugin.
+	Version string `yaml:"version"`
+
+	// Usage provides a short description
+	// of what the plugin does
+	Usage string `yaml:"usage"`
+
+	// Description provides a long description
+	// of what the plugin does
+	Description string `yaml:"description"`
+
+	// Command is the command to add to conftest
+	Command Command `yaml:"command"`
+}
+
+// Plugin represents a conftest plugin
+type Plugin struct {
+	// Metadata contains the contents of the plugin.yaml metatdata file
+	MetaData *MetaData
+
+	// Dir contains the full path to the plugin
+	Dir string
+
+	stdIn  io.Reader
+	stdOut io.Writer
+	stdErr io.Writer
+	env    []string
+}
+
+// SetStdIn configures where the plugin reads from
+// when the command is executed
+func (p *Plugin) SetStdIn(r io.Reader) *Plugin {
+	p.stdIn = r
+	return p
+}
+
+// SetStdOut configures to where the plugin writes to
+// when the command is executed
+func (p *Plugin) SetStdOut(w io.Writer) *Plugin {
+	p.stdOut = w
+	return p
+}
+
+// SetStdErr configures to where the plugin writes errors to
+// when the command is executed
+func (p *Plugin) SetStdErr(w io.Writer) *Plugin {
+	p.stdErr = w
+	return p
+}
+
+// SetEnv configures the available environment for the plugin
+func (p *Plugin) SetEnv(env []string) *Plugin {
```

Comment:
looks like it's created but not used 
what's the rationale of creating this?

Reply:
it is indeed not need, I went for a different direction with the plugin system initially.